tests/util.py

```diff
@@ -0,0 +1,171 @@
+                    yield bigquery.ArrayQueryParameter.from_api_repr(param)
+                except KeyError:
+                    # this is a different format for scalar param than above
+                    yield bigquery.ScalarQueryParameter.from_api_repr(param)
+
+
+def generate_tests():
+    tests_dir = os.path.dirname(__file__)
+    sql_dir = os.path.join(os.path.dirname(tests_dir), "sql")
+
+    # iterate over directories in tests_dir
+    for query_name in next(os.walk(tests_dir))[1]:
+        query_dir = os.path.join(tests_dir, query_name)
+
+        # read query or skip
+        try:
+            query = read(sql_dir, f"{query_name}.sql")
+        except FileNotFoundError:
+            continue
+
+        # generate a test for each directory in query_dir
+        for test_name in next(os.walk(query_dir))[1]:
+            resource_dir = os.path.join(query_dir, test_name)
+            query_params = get_query_params(resource_dir)
+            tables: Set[Table] = set()
+            replace: Dict[str, str] = {}
+
+            # load expect or skip
+            try:
+                expect = load(resource_dir, "expect")
+            except FileNotFoundError:
+                continue
+
+            # generate tables for files with a supported table extension
+            for resource in next(os.walk(resource_dir))[2]:
+                if resource in ("expect.ndjson", "query_params.ndjson"):
+                    continue  # expect and query_params are not tables
+                table_name, extension = resource, ""
+                if "." in resource:
+                    table_name, extension = resource.rsplit(".", 1)
+                if extension in table_extensions:
+                    source_format = table_extensions[extension]
+                    source_path = os.path.join(resource_dir, resource)
+                    if "." in table_name:
+                        # define replace to remove dataset from table_name in sql
+                        replace[table_name] = table_name.rsplit(".", 1)[1]
+                        # remove dataset from table_name
+                        table_name = replace[table_name]
+                    tables.add(Table(table_name, source_format, source_path))
+
+            # yield a test
+            yield GeneratedTest(
+                expect=expect,
+                name=test_name,
+                query=query,
+                query_name=query_name,
+                query_params=query_params,
+                replace=replace,
+                tables=tables,
+            )
+
+
+def coerce_result(*elements: Any):
+    for element in elements:
+        if isinstance(element, (dict, bigquery.Row)):
+            yield {
+                key: list(coerce_result(*value))
+                if isinstance(value, list)
+                else next(coerce_result(value))
+                for key, value in element.items()
+                # drop generated_time column
+                if key not in ("generated_time",)
```

Comment:
`generated_time` is currently only validated by the destination table schema if it was loaded (so for this test it's required, a datetime, and in the correct column order). we could go further and assert that it's between the start and end of the test, and maybe the same for all output rows? I'm not sure if that's worth adding to this PR

Reply:
I'm fine cutting this out of scope for this PR.

Reply:
I'll make sure to thoroughly explain it in `tests/README.md`. I might add a kwarg later to control this, or do something with schema field descriptions indicating what to drop for expect.